FiniteStateMachine/src/statemachine/StateMachine.java

```diff
@@ -3,18 +3,27 @@
 import org.jetbrains.annotations.NotNull;
 import statemachine.exceptions.BadTransitionException;
 
-import java.util.EnumSet;
+import java.util.Set;
 
+import static statemachine.exceptions.BadTransitionException.badTransitionException;
 
-public class StateContainer<E extends Enum<E> & State<E>> implements State<E> {
+
+/**
+ * A machine that has states of type {@code E} and knows its current state.
+ * It can transition between states, and will throw a {@link BadTransitionException}
+ * if an invalid transition is attempted.
+ */
+public class StateMachine<E extends State<E>> {
```

Comment:
That interface could be renamed, but I like having the StateMachine provide the same hooks as the States, so as to be able to report on itself in a consistent manner. It's also why there was a `getState` method.